Entity/CustomItem.php

```diff
@@ -611,4 +620,39 @@ public function getRelationsByType(string $entityType): Collection
                 return new ArrayCollection([]);
         }
     }
+
+    public function getUniqueHash(): ?string
+    {
+        return $this->uniqueHash;
+    }
+
+    public function setUniqueHash(?string $uniqueHash): void
+    {
+        $this->uniqueHash = $uniqueHash;
+    }
+
+    public function createUniqueHash(): ?string
+    {
+        $rowData                = [];
+        $uniqueHash             = [];
+        $uniqueIdentifierFields = $this->customObject->getFieldsIsUniqueIdentifier();
+
+        if (is_null($uniqueIdentifierFields)) {
+            return null;
+        }
+
+        foreach ($this->getCustomObject()->getCustomFields()->getValues() as $customField) {
+            try {
+                $rowData[$customField->getAlias()] = $this->findCustomFieldValueForFieldAlias($customField->getAlias())->getValue();
+            } catch (NotFoundException $exception) {
+            }
+        }
+
+        foreach ($uniqueIdentifierFields as $uniqueIdentifierField) {
+            $uniqueHash = array_merge($uniqueHash, [$uniqueIdentifierField->getAlias() => $rowData[$uniqueIdentifierField->getAlias()]]);
+        }
+        ksort($uniqueHash); //sort array on the basis of key so that the order of keys is the same everytime
+
+        return [] == $uniqueHash ? null : hash('sha256', serialize($uniqueHash));
```

Comment:
I'd suggest using json_encode. Here are some reasons why: 

https://github.com/mautic/mautic/pull/2040

But mostly because `serialize` can be broken with some UTF8 characters like emoji so it's better to avoid using it.

```suggestion
        return [] == $uniqueHash ? null : hash('sha256', json_encode($uniqueHash));
```

Reply:
This was really helpful, thank you!